logger: support log control in admin interface and command line option for Fancy Logger

PrinceS17 opened the pull request:
Commit Message: Add log control (list and modify log level) in admin interface for Fancy Logger, a new fine-grained logger for Envoy, and provide command line option --enable-fine-grain-logging for developers. 

Additional Description: A doc of overview is provided here: source/docs/fancy_logger.md.
Risk Level: Medium
Testing: Unit tests.
Docs Changes: Added a new option --enable-fine-grain-logging and doc it.
Release Notes: Added to current.rst.

lizan commented:
Can we do this switch via a command line flag rather than a compile time flag? We've done a couple of migration (for example symbol table) via command line flag, which is much easier for users to get old behavior if things break.

PrinceS17 commented:
> Can we do this switch via a command line flag rather than a compile time flag? We've done a couple of migration (for example symbol table) via command line flag, which is much easier for users to get old behavior if things break.

We thought about it and it seems not easy as all log macros are defined at compile time. My design is to replace ENVOY_LOG with FANCY_LOG so it needs compile time flag..

penguingao commented on source/common/common/fancy_logger.cc:
```diff
@@ -26,11 +26,20 @@ class FancyBasicLockable : public Thread::BasicLockable {
 void FancyContext::initFancyLogger(std::string key, std::atomic<spdlog::logger*>& logger)
     ABSL_LOCKS_EXCLUDED(fancy_log_lock_) {
+  if (LOGGER_MODE && Logger::Context::getLoggerMode() == Logger::LoggerMode::Envoy) {
+    // Enabled by compile option, but using FANCY_LOG w/o mode being set is also possible
```
How is this possible, can you elaborate more?

penguingao commented on source/common/common/logger.h:
```diff
@@ -225,25 +226,46 @@ enum class LoggerMode { Envoy, Fancy };
  * and logging will remain unlocked, the same state it is in prior to
  * instantiating a Context.
+ *
+ * Settings for Fancy Logger, a file level logger without explicit implementation,
```
explicit implementation of Envoy::Logger::Loggable.

penguingao commented on source/docs/fancy_logger.md:
````diff
@@ -0,0 +1,64 @@
+```
+the mode is **Fancy mode**. Fancy Logger is enabled and most Envoy's log macros (`ENVOY_LOG, ENVOY_FLUSH_LOG, ENVOY_CONN_LOG, ENVOY_STREAM_LOG`) are expanded to corresponding macros of Fancy Logger. In this mode, the change is transparent and developers can continue to use `ENVOY_LOG`. If logging context is defined by a main program (which is the common case), the given default log format and log ...
+
+Note that Envoy's logger can also be used in Fancy mode. These macros are not replaced: `GET_MISC_LOGGER, ENVOY_LOG_MISC, ENVOY_LOGGER, ENVOY_LOG_TO_LOGGER, ENVOY_CONN_LOG_TO_LOGGER, ENVOY_STREAM_LOG_TO_LOGGER`. For example, `ENVOY_LOG_LOGGER(ENVOY_LOGGER(), LEVEL, ...)` is equivalent to `ENVOY_LOG` in Envoy mode. 
````
s/also/still

penguingao commented on source/docs/fancy_logger.md:
```diff
@@ -0,0 +1,64 @@
+
+Note that Envoy's logger can also be used in Fancy mode. These macros are not replaced: `GET_MISC_LOGGER, ENVOY_LOG_MISC, ENVOY_LOGGER, ENVOY_LOG_TO_LOGGER, ENVOY_CONN_LOG_TO_LOGGER, ENVOY_STREAM_LOG_TO_LOGGER`. For example, `ENVOY_LOG_LOGGER(ENVOY_LOGGER(), LEVEL, ...)` is equivalent to `ENVOY_LOG` in Envoy mode. 
+
+If symbol `FANCY` is not defined, it's **Envoy mode** with existing Envoy's logger being used. In this mode, basic macros like `FANCY_LOG` can be used but the main part of `ENVOY_LOG` will keep the same. One limitation is that logger update in admin page is not supported by default as it detects Envoy mode. The reason is: Envoy mode is designed only to be back compatible. To address it, developer ...
```
macro?

PrinceS17 replied:
Sorry I don't get you...

penguingao replied:
Replace "symbol" with "macro"

PrinceS17 replied:
Okay, I see.

penguingao commented on source/docs/fancy_logger.md:
```diff
@@ -0,0 +1,64 @@
+### Runtime Update
+Runtime update of Fancy Logger is supported with administration interface, i.e. admin page, and Fancy mode needs to be enabled to use it. Same as Envoy's logger, the following functionalities are provided:
+
+1. `POST /logging`: List all active loggers and their levels;
```
Is GET supported too?

PrinceS17 replied:
GET is supported for general operations but not implemented for logging now. Do we need to support it too?

penguingao replied:
No. Just curious. Whatever the current api support is fine.

penguingao commented on source/docs/fancy_logger.md:
```diff
@@ -0,0 +1,64 @@
+
+### Implementation Details
+Fancy Logger can be divided into two parts: 
+1. Core part: file level logger without explicit inheriting `Loggable`;
```
Envoy::Logger::Loggable to more specific.

penguingao commented on test/common/common/log_macros_test.cc:
```diff
@@ -155,30 +157,62 @@ TEST(Fancy, Global) {
-  EXPECT_EQ(Logger::Context::getFancyDefaultLevel(), spdlog::level::info);
+TEST(Fancy, Iteration) {
+  FANCY_LOG(info, "Info: iteration test begins.");
+  FANCY_LOG(info, getFancyContext().listFancyLoggers());
```
Can you actually compare the return value to some expectation?

lizan commented:
If you be able to define another macro like `ENVOY_LOG(...) if (use_fancy_logger_) { FANCY_LOG(...) } else { OLD_ENVOY_LOG(...); }` to achieve command line flag, no?

penguingao commented:
> If you be able to define another macro like `ENVOY_LOG(...) if (use_fancy_logger_) { FANCY_LOG(...) } else { OLD_ENVOY_LOG(...); }` to achieve command line flag, no?

I think the problem of an if branch is that OLD_ENVOY_LOG would still be compiled which makes a compile error when extensions try to use ENVOY_LOG without explicitly inherit Loggable, especially closed source extensions.

PrinceS17 commented on source/common/common/logger.cc:
```diff
@@ -109,6 +109,8 @@ DelegatingLogSinkSharedPtr DelegatingLogSink::init() {
 
 static Context* current_context = nullptr;
 
+int Context::logger_mode_ = 0; // 0 for Envoy, 1 for Fancy
```
Because I want to use a static variable here so the logger mode exists no matter Context instance exists or not.

penguingao replied:
LoggerMode is an enum, which should work the same way as an int.

PrinceS17 commented:
> > If you be able to define another macro like `ENVOY_LOG(...) if (use_fancy_logger_) { FANCY_LOG(...) } else { OLD_ENVOY_LOG(...); }` to achieve command line flag, no?
> 
> I think the problem of an if branch is that OLD_ENVOY_LOG would still be compiled which makes a compile error when extensions try to use ENVOY_LOG without explicitly inherit Loggable, especially closed source extensions.

If we use current design, it seems that ENVOY_LOG w/o Loggable cannot be allowed (which is feasible as the Envoy mode is default and ENVOY_LOG w/o Loggable won't pass tests). Therefore, we can recommend the usage of explicit FANCY_LOG with setLoggerMode() for admin support. And at some point replace all FANCY_LOG with ENVOY_LOG.

Seems we need to check if Loggable or __log_do_not_use_read_comment() exists in the scope where ENVOY_LOG is used, but I don't know how to do it yet...

penguingao commented:
> > > If you be able to define another macro like `ENVOY_LOG(...) if (use_fancy_logger_) { FANCY_LOG(...) } else { OLD_ENVOY_LOG(...); }` to achieve command line flag, no?
> > 
> > 
> > I think the problem of an if branch is that OLD_ENVOY_LOG would still be compiled which makes a compile error when extensions try to use ENVOY_LOG without explicitly inherit Loggable, especially closed source extensions.
> 
> If we use current design, it seems that ENVOY_LOG w/o Loggable cannot be allowed (which is feasible as the Envoy mode is default and ENVOY_LOG w/o Loggable won't pass tests). Therefore, we can recommend the usage of explicit FANCY_LOG with setLoggerMode() for admin support. And at some point replace all FANCY_LOG with ENVOY_LOG.
> 
> Seems we need to check if Loggable or __log_do_not_use_read_comment() exists in the scope where ENVOY_LOG is used, but I don't know how to do it yet...

Take a look at source/server/options_impl.[h|cc] for how to use command line flags.

PrinceS17 commented:
@lizan @penguingao Just completed the command line option part by adding a new option in server_info.proto and implementing it. Take a look:)

penguingao commented on source/common/common/logger.cc:
```diff
@@ -109,10 +109,12 @@ DelegatingLogSinkSharedPtr DelegatingLogSink::init() {
+
 Context::Context(spdlog::level::level_enum log_level, const std::string& log_format,
-                 Thread::BasicLockable& lock, bool should_escape)
+                 Thread::BasicLockable& lock, bool should_escape, bool log_mode)
```
same here: use an enum or name it enable_fancy_log.

penguingao commented on source/common/common/logger.h:
```diff
@@ -385,9 +401,47 @@ template <Id id> class Loggable {
+/**
+ * Command line options for log macros: use Fancy Logger or not.
+ */
+#define LOGGER_MODE Logger::Context::useFancyLogger()
```
Is this still necessary? Can you use Logger::Context::useFancyLogger directly?

penguingao commented on source/common/common/logger.cc:
```diff
@@ -126,15 +128,33 @@ Context::~Context() {
+  }
+}
+
+LoggerMode Context::getLoggerMode() { return static_cast<LoggerMode>(logger_mode_); }
```
Why do you need both a static logger_mode_ and a member log_mode_?

I'd imagine you have just a getLoggerMode() that returns current_context->log_mode_, which should be an enum.

lizan commented on source/docs/fancy_logger.md:
````diff
@@ -0,0 +1,63 @@
+```
+To flush a logger, `FANCY_FLUSH_LOG()` can be used. 
+
+### Logger Mode: Fancy or Envoy [Deprecated: commandline option now]
````
We don't have any usage of this, just delete this paragraph and let folks use command line option?

PrinceS17 replied:
Addressed now. I think the paragraph should be kept as there's still something to remind user.

lizan commented on source/docs/fancy_logger.md:
```diff
@@ -0,0 +1,63 @@
+Runtime update of Fancy Logger is supported with administration interface, i.e. admin page, and Fancy mode needs to be enabled to use it. Same as Envoy's logger, the following functionalities are provided:
+
+1. `POST /logging`: List all active loggers and their levels;
+2. `POST /logging?<file_path>=<level>`: Given path from Envoy's root directory, change the log level of the file;
```
We probably need a better handling of file_path, it can be leaved as a follow up but FWIW:

- Header file path are not deterministic and depends on build configuration (bazel-out/k8-{fastbuild,opt,dbg}/blah)
- Downstream builds will have different path prefix like (`external/envoy/source/common/common/logger.cc`)

A naive solution would be doing a suffix match or regex match

PrinceS17 replied:
I think current method is straightforward and the user can list loggers to make sure about their input, but yes, the handle of file path can be definitely improved in the future.

lizan replied:
That doesn't match the description here, it's not path from Envoy's root directory for some files, and not true for all files in downstream builds.

PrinceS17 replied:
Ah you are right. I just modified the doc.

PrinceS17 commented on source/docs/fancy_logger.md:
````diff
@@ -0,0 +1,64 @@
+```
+If the level of log message is higher than that of the file, macros above will print messages with the file name like this:
+```
+[2020-07-29 22:27:02.594][15][error][test/common/common/log_macros_test.cc] [test/common/common/log_macros_test.cc:149] Fancy Error! Here\'s the second message!
````
Addressed: now file name (i.e. logger name in Fancy Logger) is excluded if default format is used and it's Fancy Logger mode. I think it should depend on user if the format is manually set so logger name isn't excluded there.

htuch commented:
Drive-by: are we really going to call this fancy logger? I think we should maybe do something a bit more descriptive, e.g. "fine-grained logger".

penguingao commented:
LGTM. @htuch re naming: I agree with a better naming, but it's Jinhui's last week and he has been using FancyLogger. I'd suggest we name the command line options --enable-fine-grain-log and the associated proto field enable_fine_grain_log for now so that the API is stable while I can come back and clean the internal naming afterwards.

Would that be okay?

htuch commented:
@penguingao sounds like a plan, with the tiny nit that I'd prefer `enable_fine_grain_logging` rather than `enable_fine_grain_log` etc. Basically, just make sure the user/API visible bits don't reference fancy. This is really nice work BTW.

penguingao commented:
Could you update the PR description to reflect the command line flag instead of describing the compile option?

PrinceS17 commented:
> Could you update the PR description to reflect the command line flag instead of describing the compile option?

Definitely, in a second.

PrinceS17 commented:
@lizan @penguingao @htuch @envoyproxy/api-shepherds
Finally I addressed most reviews for current version and got it pass the tests:) The file path issue is left for future improvement and the inside variable name of Fancy can also be replaced in the future. Could you guys give a look to see if we can get it approved? Thanks!

htuch commented:
/lgtm api
